Media Type & <RichMedia />

rgbkrk opened the pull request:
~~Super WIP here! Just posting so everyone knows which segment I'm touching.~~

Taking a stab at a new interface for the `<ExecuteResult />`, `<DisplayData />`, and `<RichMedia />`. This starts with _only_ `<RichMedia />` for the moment.

Demo: https://declarative-trial.nteract.io/#!/RichMedia

Here's a snippet of the new interface:

```jsx
// Declare a component for rendering plaintext
const Plain = props => <pre>{props.data}</pre>;
Plain.defaultProps = {
  mediaType: "text/plain"
};

// Declare a component for rendering HTML
const HTML = props => <div dangerouslySetInnerHTML={{ __html: props.data }} />;
HTML.defaultProps = {
  mediaType: "text/html"
};

<RichMedia
  data={{
    "text/plain": "plain was richer",
    "text/html": "<b>HTML is so much richer</b>"
  }}
>
  <HTML />
  <Plain />
</RichMedia>
```


Miscellaneous TODO as I come across them, so I don't forget:

* ~~`@nteract/presentational-components` now depends on `@nteract/records` solely for types -- need to figure out if this should be a hard dependency~~ included a simple type for now...

For later:

* Much of the styling of `<Outputs />` should only be affecting the html and markdown types. Right now it's this global grab bag of things.
* ~~We need to make it super easy to special case handling of the data explorer, to give it control over height~~ Hopefully this is good now with our declarative `<RichMedia>` component
* ~~How we do transforms should become simpler _somehow_~~ Possibly solved here
* ~~`onMetadataChange` should be passed in here~~ Hey I think we can at least do this with the Data Explorer with this setup.

rgbkrk commented:
> `@nteract/presentational-components` now depends on `@nteract/records` solely for types -- need to figure out if this should be a hard dependency

I now realize this makes documenting the flow types weird so I think I'll write them again here so that I can declare what the props mean in our components.

rgbkrk commented:
Ok, I'm loving this interface experiment. It rethinks how we've been doing `displayOrder` and `transforms` for mimebundle outputs.

Play with it now via https://nteract-components-yizfzoaduc.now.sh/#!/ExecuteResult

As a preview, you declare which components you'll render with as children, in order of decreasing priority:

```
<ExecuteResult
  data={{
    "text/plain": "plain was richer",
    "text/html": "<b>HTML was richer</b>"
  }}
>
  <HTML />
  <Plain />
</ExecuteResult>;
```

Any props that get passed in to these children get passed on through when rendered so you can pass in custom things:

```
<ExecuteResult
  data={{
    "text/plain": "plain was richer",
    "text/html": "<b>HTML was richer</b>"
  }}
>
  <HTML iframeMe="yesplease" />
  <Plain />
</ExecuteResult>;
```

🎉 

In terms of how our current interface works, we can adapt old code like this:

```
<ExecuteResult
  data={{
    "text/plain": "plain was richer",
    "text/html": "<b>HTML was richer</b>"
  }}
>
  {displayOrder.map(mimetype => transforms[mimetype])}
</ExecuteResult>
```

alexandercbooth commented on packages/presentational-components/src/components/execute-result.js:
```diff
@@ -0,0 +1,74 @@
+      metadata: this.props.metadata[mimetype]
+    });
+  }
+}
```
I love this! It is so much easier to understand this way 👏.

BenRussert commented:
Should we start a dev branch as this unfolds instead of waiting until we can merge into master?

rgbkrk commented on packages/presentational-components/src/components/execute-result.js:
```diff
@@ -0,0 +1,74 @@
+    });
+
+    if (chosenOne === null) {
+      return null;
```
I think `null` is the way to go. People sometimes emit mimetypes that aren't for rendering and are just for forming records. Papermill does this.

rgbkrk commented on packages/presentational-components/src/components/execute-result.md:
````diff
@@ -0,0 +1,185 @@
+}
+```
+
+The standard practice in Jupyter apps is to pick the "richest" of these. This Component lets
````
Oh excellent, I'll incorporate this today.

rgbkrk commented:
I just made some changes while traveling and will incorporate feedback later today. Thanks all for the feedback, this is great!

> Should we start a dev branch as this unfolds instead of waiting until we can merge into master?

I think we'll be able to merge this tonight without worrying about it interfering with the main app since it's a new interface. 😊

rgbkrk commented on packages/records/src/outputs/execute-result.js:
```diff
@@ -22,7 +22,7 @@ export type ExecutionCount = ?number;
 // In-memory version
-export type ExecuteResult = {
+export type ExecuteResultOutput = {
   outputType: ExecuteResultType,
```
Yeah, especially since this is an output type to match the jupyter type.

rgbkrk commented:
Ok I've updated a lot of the language in the component docs and switched just about everything over to `Media*` style naming. Ready for next round of reviews (annnnd maybe a merge ✅  😄 ).

alexandercbooth commented:
Being able to review things with accompanying styleguidist entries rocks 😄

rgbkrk commented:
Ok those were my last last last changes. I'm off to bed. 😅  🛏
